components/places/src/bookmark_sync/store.rs

```diff
@@ -391,9 +409,117 @@ impl<'a> BookmarksStore<'a> {
         Ok(())
     }
 
-    fn maybe_store_url(&self, url: Option<&String>) -> Result<Url> {
+    fn store_incoming_tombstone(&self, modified: ServerTimestamp, guid: &SyncGuid) -> Result<()> {
+        self.db.execute_named_cached(
+            r#"REPLACE INTO moz_bookmarks_synced(guid, parentGuid, serverModified, needsMerge,
+                                                 dateAdded)
+               VALUES(:guid, NULL, :serverModified, 1, 0)"#,
+            &[
+                (":guid", guid),
+                (":serverModified", &(modified.as_millis() as i64)),
+            ],
+        )?;
+        Ok(())
+    }
+
+    fn determine_query_url_and_validity(
+        &self,
+        q: &QueryRecord,
+        url: Url,
+    ) -> Result<(Option<Url>, SyncedBookmarkValidity)> {
+        // wow - this  is complex, but markh is struggling to see how to
+        // improve it
```

Comment:
We could return `Result<(Url, SyncedBookmarkValidity)>` (and an error for invalid URLs), and have `store_incoming_query` assume than error means `Replace`, but that doesn't help flatten or simplify this.